plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java

```diff
@@ -188,6 +192,50 @@ public FileHiveMetastore(NodeVersion nodeVersion, HdfsEnvironment hdfsEnvironmen
         }
     }
 
+    @Override
+    public synchronized MetastoreLock acquireTableExclusiveLock(
+            HiveIdentity identity,
+            AcidTransactionOwner transactionOwner,
+            String queryId,
+            String dbName,
+            String tableName)
+    {
+        SchemaTableName schemaTableName = new SchemaTableName(dbName, tableName);
+
+        ReentrantLock lock = tableLocks.computeIfAbsent(schemaTableName, ignored -> new ReentrantLock());
+        lock.lock();
+
+        tableLocks.put(schemaTableName, lock);
+        return new FileMetastoreLock(schemaTableName);
+    }
+
+    private class FileMetastoreLock
+            implements MetastoreLock
+    {
+        private final SchemaTableName tableName;
+        boolean released;
+
+        public FileMetastoreLock(SchemaTableName tableName)
+        {
+            this.tableName = requireNonNull(tableName, "tableName is null");
+        }
+
+        @Override
+        public void release()
+        {
+            if (released) {
+                throw new IllegalStateException("Table lock has already been released");
+            }
+            released = true;
+            ReentrantLock lock = tableLocks.get(tableName);
+            // Currently, metastore locks are always acquired and released in the same thread.
+            lock.unlock();
+            if (!lock.isLocked()) {
```

Comment:
when can `lock.isLocked()` be true?
the locks are not supposed to use re-entrantly, i.e. the caller shouldn't acquire a given lock more than once, should they?

let's have a `verify` in acquire if that's indeed never the case
let's have a comment here if that can be the case.

Reply:
If a table lock is requested but the table is already locked, the calling thread waits until it is released and then immediately acquires it. 

If we'd rather throw an exception in that case this code could be a lot simpler